src/sender_queue/mod.rs

```diff
@@ -220,9 +238,20 @@ where
             return Step::<D>::default();
         }
         // Look up `DynamicHoneyBadger` epoch updates and collect any added peers.
-        for node in step.output.iter().flat_map(|batch| batch.added_peers()) {
-            if &node != self.our_id() {
-                self.peer_epochs.entry(node).or_default();
+        for batch in &step.output {
+            if let Some((current_participants, next_participants)) = batch.participant_transition()
+            {
+                // Insert candidates.
+                for id in &next_participants {
+                    if id != self.our_id() {
+                        self.peer_epochs.entry(id.clone()).or_default();
+                    }
+                }
+                // Remove obsolete participants.
+                for id in current_participants.difference(&next_participants) {
```

Comment:
Could we instead take the difference of `peer_epochs.keys()` and `next_participants` (as sets, or using `filter`)? Then we wouldn't need `current_participants` at all, and would correctly handle the case of two subsequent `InProgress` after a failed key generation attempt.

Reply:
No. See above. We should take the difference of current and next participants.

Reply:
Sorry, I might be a bit confused by all the sets… :dizzy_face: 
What I mean is: If `participant_transition` simply returned a single list, namely the IDs of _all_ future participants, i.e. all nodes that will be either a validator or a candidate in the next epoch, then I imagine that would suffice? Because then the difference of `peer_epochs.keys()` and `next_participants` would be exactly the peers that won't participate (either as candidates or as validators) in the next epoch anymore?